server/automute.go

```diff
@@ -0,0 +1,214 @@
+package main
+
+import (
+	"database/sql"
+	"fmt"
+	"strconv"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/pkg/errors"
+)
+
+const (
+	NotifyPropAutomuted = "msteams_automute"
+
+	// This preference tracks if we've enabled automuting for the user by muting all of their connected channels.
+	PreferenceNameAutomuteEnabled = "automute_enabled"
+)
+
+func (p *Plugin) tryEnableAutomute(userID string, skipConnectedCheck, skipPrimaryPlatformCheck bool) (bool, error) {
+	enabled, err := p.shouldEnableAutomuteForUser(userID, skipConnectedCheck, skipPrimaryPlatformCheck)
+	if err != nil {
+		return false, err
+	} else if !enabled {
+		return false, nil
+	}
+
+	return p.setAutomuteEnabledForUser(userID, true)
+}
+
+// shouldEnableAutomuteForUser returns true if the given user is both connected to Teams and has their primary platform
+// set to Teams.
+//
+// skipConnectedCheck and skipPrimaryPlatformCheck can be used to skip the respective checks if the calling code has
+// already performed those checks.
+func (p *Plugin) shouldEnableAutomuteForUser(userID string, skipConnectedCheck, skipPrimaryPlatformCheck bool) (bool, error) {
+	var connected bool
+	if skipConnectedCheck {
+		connected = true
+	} else {
+		var err error
+		connected, err = p.isUserConnected(userID)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	var teamsPrimary bool
+	if skipPrimaryPlatformCheck {
+		teamsPrimary = true
+	} else {
+		teamsPrimary = p.isUsersPrimaryPlatformTeams(userID)
+	}
+
+	if !connected || !teamsPrimary {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (p *Plugin) tryDisableAutomute(userID string) (bool, error) {
+	return p.setAutomuteEnabledForUser(userID, false)
+}
+
+func (p *Plugin) setAutomuteEnabledForUser(userID string, automuteEnabled bool) (bool, error) {
+	if channelsAutomuted := p.getAutomuteIsEnabledForUser(userID); channelsAutomuted == automuteEnabled {
+		// We've already automuted all the users' channels, so there's nothing to do
+		return false, nil
+	}
+
+	var membersToMute []*model.ChannelMemberIdentifier
+
+	if channels, appErr := p.API.GetChannelsForTeamForUser("", userID, true); appErr != nil {
+		return false, errors.Wrap(appErr, fmt.Sprintf("Unable to get channels for user %s to automute them", userID))
+	} else {
+		for _, channel := range channels {
+			if linked, err := p.canAutomuteChannel(channel); err != nil {
+				return false, err
+			} else if !linked {
+				continue
+			}
+
+			membersToMute = append(membersToMute, &model.ChannelMemberIdentifier{ChannelId: channel.Id, UserId: userID})
+		}
+	}
+
+	if err := p.setChannelMembersAutomuted(membersToMute, automuteEnabled); err != nil {
+		return false, errors.Wrap(err, fmt.Sprintf("Unable to automute all channels for linked user %s", userID))
+	}
+
+	if err := p.setAutomuteIsEnabledForUser(userID, automuteEnabled); err != nil {
+		return false, err
+	}
```

Comment:
Any potential race condition because of doing this at the end? Hypothetical scenario that comes to mind:
- Change from false to true. The process (A) start.
- Change back to false. The new process (B) will stop at `p.getAutomuteIsEnabledForUser`
- (A) finishes, and set this to true, even if the user thought they put it to false

May be too corner case to consider, but wanted to raise it either way.

Reply:
I have been thinking that moving this piece of code somewhere else would only create a different race condition (potentially worse). If we were to address this at any point, we would have to use some kind of HA mutex (write into the KV store, for example). And even then, we would still have problems :sweat_smile:

Reply:
There's some potential raciness here as well if the user joins/leaves a channel at the same time this logic is running. Using a transaction or a more purpose-built SQL query could help that, but it'd definitely need some sort of locking either on the app level or DB level.

It does seem like something good to be aware of, but like you said, I'm not sure how possible it would be to prevent this even with something really complicated. At least this is just about muting and isn't super critical feature, so I think it's okay even if it's not 100% correct in weird cases 😅